tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_Hierarchical.cpp

```diff
@@ -0,0 +1,2383 @@
+  helper({7.0, 5.0, 2.5, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 1.1, 1.0e-18, 0.0},
+         {3.0, 2.0, 0.9 * 0.99, 0.5 * 0.99});
+  // Limit top upper neighbor, zero second upper neighbor
+  helper({7.0, 1.0, 2.5, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 1.1, 1.0e-18, 0.0},
+         {3.0, 2.0, 0.0, 0.5 * 0.99});
+  // Limit top upper neighbor, zero second lower neighbor
+  helper({7.0, 5.0, 2.5, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 2.1, 1.0e-18, 0.0},
+         {3.0, 2.0, 0.0, 0.5 * 0.99});
+  // Zero top upper neighbor, limit second upper neighbor
+  helper({7.0, 3.0, 1.9, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, -5.0, 1.0e-18, 0.0},
+         {3.0, 2.0, 1.0 * 0.99, 0.0});
+  // Zero top upper neighbor, limit second lower neighbor
+  helper({7.0, 5.0, 1.9, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 1.1, 1.0e-18, 0.0},
+         {3.0, 2.0, 0.9 * 0.99, 0.0});
+  // Zero top upper neighbor, zero second upper neighbor
+  helper({7.0, 1.0, 1.9, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 1.1, 1.0e-18, 0.0},
+         {3.0, 2.0, 0.0, 0.0});
+  // Zero top upper neighbor, zero second lower neighbor
+  helper({7.0, 5.0, 1.9, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 2.1, 1.0e-18, 0.0},
+         {3.0, 2.0, 0.0, 0.0});
+
+  // Limit top lower neighbor, limit second upper neighbor
+  helper({7.0, 3.0, 4.1, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, -5.0, 1.3, 0.0},
+         {3.0, 2.0, 1.0 * 0.99, 0.7 * 0.99});
+  // Limit top lower neighbor, limit second lower neighbor
+  helper({7.0, 5.0, 4.1, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 1.1, 1.3, 0.0},
+         {3.0, 2.0, 0.9 * 0.99, 0.7 * 0.99});
+  // Limit top lower neighbor, zero second upper neighbor
+  helper({7.0, 1.0, 4.1, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 1.1, 1.3, 0.0},
+         {3.0, 2.0, 0.0, 0.7 * 0.99});
+  // Limit top lower neighbor, zero second lower neighbor
+  helper({7.0, 5.0, 4.1, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 2.1, 1.3, 0.0},
+         {3.0, 2.0, 0.0, 0.7 * 0.99});
+  // Zero top lower neighbor, limit second upper neighbor
+  helper({7.0, 3.0, 4.1, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, -5.0, 3.0, 0.0},
+         {3.0, 2.0, 1.0 * 0.99, 0.0});
+  // Zero top lower neighbor, limit second lower neighbor
+  helper({7.0, 5.0, 4.1, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 1.1, 3.0, 0.0},
+         {3.0, 2.0, 0.9 * 0.99, 0.0});
+  // Zero top lower neighbor, zero second upper neighbor
+  helper({7.0, 1.0, 4.1, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 1.1, 3.0, 0.0},
+         {3.0, 2.0, 0.0, 0.0});
+  // Zero top lower neighbor, zero second lower neighbor
+  helper({7.0, 5.0, 4.1, 0.0}, {3.0, 2.0, 2.0, 1.0}, {-2.1, 2.1, 3.0, 0.0},
+         {3.0, 2.0, 0.0, 0.0});
+}
+
+void run() noexcept {
+  for (size_t order = 2; order < 7; ++order) {
+    package_data(order);
+  }
+  two_neighbors();
+}
+}  // namespace test_1d
+
+namespace test_2d {
+template <typename F>
+void test_2_2(const F& helper) noexcept {
+  // Limit no coefficients because the highest coefficient doesn't need
+  // limiting.
+  helper({0.0, 1.0, 2.0, 3.0, 4.0, 7.0, 3.0, 4.1, 0.0},
```

Comment:
I understand that limiters are complicated, and it's important to hit every branch, but a lot of these long inputs seem to be very similar values with a couple of small alterations to make sure the limiter is applied to certain combinations. Is there no way that these values could be produced algorithmically? It causes me anxiety to consider the possibility of someone having to update these tests if something changes.

I also understand that finding an automatic way of producing these values could be a nontrivial undertaking, so I'm happy to just make an issue for this if you agree it is something that can and should be done.

Reply:
In theory, it's probably possible, in practice it's rather difficult because there are non-trivial dependencies between all the modal coefficients. My other concern with writing an algorithm is that that algorithm is basically the inverse of the limiter and given how complicated the limiter is I don't think it's any more likely that someone will implement the inverse algorithm correctly. I think testing individual numbers is the safest thing to do.

Reply:
Understood :+1: .